### ISSUE TITLE
PLN: support going from PredicateNode to EvaluationLink

### ISSUE THREAD

**Issue by williampma** [opencog/opencog#1213]
Needed for R2L outputs like:
```
(EvaluationLink
   (PredicateNode "that")
   (ListLink
      (PredicateNode "glad@11")
      (PredicateNode "won@22")))
(EvaluationLink
   (PredicateNode "glad@11")
   (ListLink
       (ConceptNode "he@33")))
(EvaluationLink
   (PredicateNode "won@22")
   (ListLink
       (ConceptNode "she@44")))
```
see https://github.com/opencog/relex/issues/124

@AmeBel add this to the DialogueSystem milestone?

**Comment by anitzkin** [comment]
What is being added to what here? All I see above is an illustration of a very serious problem with R2L that I have been pointing out for months now . . .

**Comment by williampma** [comment]
There're all sort of problems with R2L.  I am just adding this issue so we do not forget it (and because I happen to come across this again while working with SuReal)

**Comment by anitzkin** [comment]
Ah. yeah.

**Comment by anitzkin** [comment]
I believe the solution to this problem is (in cognitive linguistics) is "thematic roles" and "thematic relations" but that implies a lot -- a lot that is almost surely going to be needed eventually.

**Comment by AmeBel** [comment]
@williampma  i don't get what PLN is supposed to do and why.

**Comment by anitzkin** [comment]
I don't know if we are talking about the same issue, but what I see is that PLN needs to determine which "thematic role" each argument of the predicate plays in the sentence.  This is necessary for the system to understand what situation is being described, and it is, in human cognition, determined by probabilistic logic.  There is a good paper on it by Leonard Talmy, that I will try to find tomorrow if you are interested . . .

**Comment by linas** [comment]
@anitzkin Yes, one does need to identify rhemes and themes but not in PLN.  PLN knows nothing about sentences or even language. The right way to think of PLN as a rather low-level rule engine.  It takes a set of graphs, re-writes them, twiddles their truth values in various ways, does some forward-chaining deduction.   Its a generic mechanism, and has nothing to do with natural-language per-se, or with understanding language.

We'll be using PLN as the tool to do the high-level reasoning about language ... but that's later..

**Comment by bgoertzel** [comment]
Aaron,

Are you  thinking of tagging each argument-slot of a predicate with an
indication of the semantic role that it's intended to fulfill?  (which
involves choosing one of the N frameworks for semantic roles out
there)....   Or something more sophisticated?

We did that in a previous OpenCog version using FrameNet's system, which
had pluses and minuses, but had the advantage of being well laid-out in
electronic form and coming along with a fairly big dictionary....

The old RelEx2Frame system was f**kedin a number of ways, but it could be
viable to reintroduce FrameNet in a more limited way in order to utilize
its system of semantic role labels...

-- Ben

On Tue, Jan 13, 2015 at 3:41 AM, anitzkin <notifications@github.com> wrote:

> I believe the solution to this problem is (in cognitive linguistics) is
> "thematic roles" and "thematic relations" but that implies a lot -- a lot
> that is almost surely going to be needed eventually.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1213#issuecomment-69718650>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by bgoertzel** [comment]
Yes, I mean assigning each argument in a specific sentence with a role such
as Agent, Patient, Theme, Landmark, etc. i wouldn't
say tagging each slot since the sentence-specific assignment of roles
requires probabilistic reasoning based on prototype-category
like criteria.  So, yes, that sounds like what you're talking about.  It
would be good to see that . . .

**Comment by anitzkin** [comment]
Well, Linas, these are exactly the kind of category assignments required to
transform 'meaningless' reasoning into
meaningful reasoning.  The thing is, that we're already doing this,
partially, inconsistently, somewhat randomly.  LG
has quite a few link types that distinguish themselves by semantic
criteria, and our relex dependency relations are
at present a jumble ranging from purely syntactic to purely semantic and
everything in between; I suggest that we
try a principled, fore-thought-out version!

**Comment by anitzkin** [comment]
Have no idea how this reply of mine to you ended up getting attributed to you, ben!

--------------------------------------------
On Wed, 1/14/15, bgoertzel <notifications@github.com> wrote:

 Subject: Re: [opencog] PLN: need support going from PredicateNode to EvaluationLink (#1213)
 To: "opencog/opencog" <opencog@noreply.github.com>
 Cc: "anitzkin" <a_nitzkin@yahoo.com>
 Date: Wednesday, January 14, 2015, 8:24 AM
 
 Yes, I mean assigning each argument
 in a specific sentence with a role such
 
 
 as Agent, Patient, Theme, Landmark, etc. i wouldn't
 
 
 say tagging each slot since the sentence-specific assignment
 of roles
 
 
 requires probabilistic reasoning based on prototype-category
 
 
 like criteria.  So, yes, that sounds like what you're
 talking about.  It
 
 
 would be good to see that . . .
 
 
 
 
 
 On Wed, Jan 14, 2015 at 2:02 AM, Ben Goertzel
 <ben@goertzel.org> wrote:
 
 
 
 
 
 >
 
 
 > Aaron,
 
 
 >
 
 
 > Are you  thinking of tagging each argument-slot of a
 predicate with an
 
 
 > indication of the semantic role that it's intended
 to fulfill?  (which
 
 
 > involves choosing one of the N frameworks for semantic
 roles out
 
 
 > there)....   Or something more sophisticated?
 
 
 >
 
 
 > We did that in a previous OpenCog version using
 FrameNet's system, which
 
 
 > had pluses and minuses, but had the advantage of being
 well laid-out in
 
 
 > electronic form and coming along with a fairly big
 dictionary....
 
 
 >
 
 
 > The old RelEx2Frame system was f**kedin a number of
 ways, but it could be
 
 
 > viable to reintroduce FrameNet in a more limited way in
 order to utilize
 
 
 > its system of semantic role labels...
 
 
 >
 
 
 > -- Ben
 
 
 >
 
 
 > On Tue, Jan 13, 2015 at 3:41 AM, anitzkin
 <notifications@github.com>
 
 
 > wrote:
 
 
 >
 
 
 >> I believe the solution to this problem is (in
 cognitive linguistics) is
 
 
 >> "thematic roles" and "thematic
 relations" but that implies a lot -- a lot
 
 
 >> that is almost surely going to be needed
 eventually.
 
 
 >>
 
 
 >> —
 
 
 >> Reply to this email directly or view it on GitHub
 
 
 >>
 <https://github.com/opencog/opencog/issues/1213#issuecomment-69718650>.
 
 
 >>
 
 
 >
 
 
 >
 
 
 >
 
 
 > --
 
 
 > Ben Goertzel, PhD
 
 
 > http://goertzel.org
 
 
 >
 
 
 > "The reasonable man adapts himself to the world:
 the unreasonable one
 
 
 > persists in trying to adapt the world to himself.
 Therefore all progress
 
 
 > depends on the unreasonable man." -- George
 Bernard Shaw
 
 
 >
 
 
 > --
 
 
 > You received this message because you are subscribed to
 the Google Groups
 
 
 > "opencog" group.
 
 
 > To unsubscribe from this group and stop receiving
 emails from it, send an
 
 
 > email to opencog+unsubscribe@googlegroups.com.
 
 
 > To post to this group, send email to
 opencog@googlegroups.com.
 
 
 > Visit this group at
 http://groups.google.com/group/opencog.
 
 
 > For more options, visit
 https://groups.google.com/d/optout.
 
 
 >
 
 —
 Reply to this email directly or view
 it on GitHub.

**Comment by bgoertzel** [comment]
So the problem with FrameNet frames such as

https://framenet2.icsi.berkeley.edu/fnReports/data/frameIndex.xml?frame=Ingestion

(the frame for the main sense of "eat" example)

is that they apply to a sense-disambiguated version
of a word...

So I really wonder if this needs to be done in a phase
*after* R2L as we're currently conceiving it.....  Without
some smart sense disambiguation, all we can do is associate
an instance of a verb with a list of *potential* semantic role labels
for its arguments...


On Tue, Jan 13, 2015 at 6:24 PM, Aaron Nitzkin <nitzkina@gmail.com> wrote:

> Yes, I mean assigning each argument in a specific sentence with a role
> such as Agent, Patient, Theme, Landmark, etc. i wouldn't
> say tagging each slot since the sentence-specific assignment of roles
> requires probabilistic reasoning based on prototype-category
> like criteria.  So, yes, that sounds like what you're talking about.  It
> would be good to see that . . .
>
> On Wed, Jan 14, 2015 at 2:02 AM, Ben Goertzel <ben@goertzel.org> wrote:
>
>>
>> Aaron,
>>
>> Are you  thinking of tagging each argument-slot of a predicate with an
>> indication of the semantic role that it's intended to fulfill?  (which
>> involves choosing one of the N frameworks for semantic roles out
>> there)....   Or something more sophisticated?
>>
>> We did that in a previous OpenCog version using FrameNet's system, which
>> had pluses and minuses, but had the advantage of being well laid-out in
>> electronic form and coming along with a fairly big dictionary....
>>
>> The old RelEx2Frame system was f**kedin a number of ways, but it could be
>> viable to reintroduce FrameNet in a more limited way in order to utilize
>> its system of semantic role labels...
>>
>> -- Ben
>>
>> On Tue, Jan 13, 2015 at 3:41 AM, anitzkin <notifications@github.com>
>> wrote:
>>
>>> I believe the solution to this problem is (in cognitive linguistics) is
>>> "thematic roles" and "thematic relations" but that implies a lot -- a lot
>>> that is almost surely going to be needed eventually.
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/opencog/opencog/issues/1213#issuecomment-69718650>.
>>>
>>
>>
>>
>> --
>> Ben Goertzel, PhD
>> http://goertzel.org
>>
>> "The reasonable man adapts himself to the world: the unreasonable one
>> persists in trying to adapt the world to himself. Therefore all progress
>> depends on the unreasonable man." -- George Bernard Shaw
>>
>> --
>> You received this message because you are subscribed to the Google Groups
>> "opencog" group.
>> To unsubscribe from this group and stop receiving emails from it, send an
>> email to opencog+unsubscribe@googlegroups.com.
>> To post to this group, send email to opencog@googlegroups.com.
>> Visit this group at http://groups.google.com/group/opencog.
>> For more options, visit https://groups.google.com/d/optout.
>>
>
>  --
> You received this message because you are subscribed to the Google Groups
> "opencog" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to opencog+unsubscribe@googlegroups.com.
> To post to this group, send email to opencog@googlegroups.com.
> Visit this group at http://groups.google.com/group/opencog.
> For more options, visit https://groups.google.com/d/optout.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by bgoertzel** [comment]
Upon reflection . . . I think most predicates could get permanent specific
role-tags on their argument slots, but not all;
For example "The robot rolled down the hill," and "The rock rolled down the
hill," show how the subject-slot of "roll"
needs to be open to being assgined an Agent or a Theme, depending on
probabilistic reasoning with prototype criteria.
Generally a lot of argument-slots can take Agent, Theme, or Patient (in
that order), depending on the animacy of the
arguments and the other roles present in the sentence.

**Comment by bgoertzel** [comment]
Aaron,

Yeah, if your semantic role labels are as broad as Agent, Theme and
Patient, then you'll find that many (though by no means all) verbs have
common labels for their argument slots, across all their different senses

To make this work we will need to use some database of semantic role labels
assigned to subcategorization frames of verbs, and have RelEx2Logic (or
some R2L postprocessor, whatever) consult this DB to assign the labels in
an automated way...

FrameNet is one such DB but not necessarily the best one, since the
strength of FrameNet is really its more refined role-labels, which
according to this plan we wouldn't be using....  If we just want simple
labels, PropBank is likely the best choice

This paper reports good results from using supervised learning on
dependency parses to assign PropBank role labels,


http://www.aclweb.org/anthology/D08-1008

with code at

http://nlp.cs.lth.se/

This could be replicated using link parses, RelEx and R2L output as
features.   This would be a pretty good GSoC project.  It's not trivial but
not really hard either, given the above paper as guidance...

Or as a nasty hack, we could use their role labeler for now, and correlate
its output with the output of our parser and R2L....  This might work just
fine in nearly all cases, though it's obviously not as clean and elegant as
replicating their idea using our own dependency parsing/interpretation
output...

-- Ben





On Tue, Jan 13, 2015 at 7:01 PM, Aaron Nitzkin <nitzkina@gmail.com> wrote:

> Upon reflection . . . I think most predicates could get permanent specific
> role-tags on their argument slots, but not all;
> For example "The robot rolled down the hill," and "The rock rolled down
> the hill," show how the subject-slot of "roll"
> needs to be open to being assgined an Agent or a Theme, depending on
> probabilistic reasoning with prototype criteria.
> Generally a lot of argument-slots can take Agent, Theme, or Patient (in
> that order), depending on the animacy of the
> arguments and the other roles present in the sentence.
>
> On Wed, Jan 14, 2015 at 8:24 AM, Aaron Nitzkin <nitzkina@gmail.com> wrote:
>
>> Yes, I mean assigning each argument in a specific sentence with a role
>> such as Agent, Patient, Theme, Landmark, etc. i wouldn't
>> say tagging each slot since the sentence-specific assignment of roles
>> requires probabilistic reasoning based on prototype-category
>> like criteria.  So, yes, that sounds like what you're talking about.  It
>> would be good to see that . . .
>>
>> On Wed, Jan 14, 2015 at 2:02 AM, Ben Goertzel <ben@goertzel.org> wrote:
>>
>>>
>>> Aaron,
>>>
>>> Are you  thinking of tagging each argument-slot of a predicate with an
>>> indication of the semantic role that it's intended to fulfill?  (which
>>> involves choosing one of the N frameworks for semantic roles out
>>> there)....   Or something more sophisticated?
>>>
>>> We did that in a previous OpenCog version using FrameNet's system, which
>>> had pluses and minuses, but had the advantage of being well laid-out in
>>> electronic form and coming along with a fairly big dictionary....
>>>
>>> The old RelEx2Frame system was f**kedin a number of ways, but it could
>>> be viable to reintroduce FrameNet in a more limited way in order to utilize
>>> its system of semantic role labels...
>>>
>>> -- Ben
>>>
>>> On Tue, Jan 13, 2015 at 3:41 AM, anitzkin <notifications@github.com>
>>> wrote:
>>>
>>>> I believe the solution to this problem is (in cognitive linguistics) is
>>>> "thematic roles" and "thematic relations" but that implies a lot -- a lot
>>>> that is almost surely going to be needed eventually.
>>>>
>>>> —
>>>> Reply to this email directly or view it on GitHub
>>>> <https://github.com/opencog/opencog/issues/1213#issuecomment-69718650>.
>>>>
>>>
>>>
>>>
>>> --
>>> Ben Goertzel, PhD
>>> http://goertzel.org
>>>
>>> "The reasonable man adapts himself to the world: the unreasonable one
>>> persists in trying to adapt the world to himself. Therefore all progress
>>> depends on the unreasonable man." -- George Bernard Shaw
>>>
>>> --
>>> You received this message because you are subscribed to the Google
>>> Groups "opencog" group.
>>> To unsubscribe from this group and stop receiving emails from it, send
>>> an email to opencog+unsubscribe@googlegroups.com.
>>> To post to this group, send email to opencog@googlegroups.com.
>>> Visit this group at http://groups.google.com/group/opencog.
>>> For more options, visit https://groups.google.com/d/optout.
>>>
>>
>>
>  --
> You received this message because you are subscribed to the Google Groups
> "opencog" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to opencog+unsubscribe@googlegroups.com.
> To post to this group, send email to opencog@googlegroups.com.
> Visit this group at http://groups.google.com/group/opencog.
> For more options, visit https://groups.google.com/d/optout.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by williampma** [comment]
Well... that certainly ended up sounding complicated.  What I had envisioned originally was just some basic predicate logic (correct or not, not sure), and has nothing to do with language even.  Everytime a `PredicateNode` appears **not** as the first argument of an `EvaluationLink`, PLN could apply some inference to replace it with an `EvaluatoinLink` where the `PredicateNode` **is** the first argument of an `EvaluationLink`.  So, with the example in the original post, PLN would end up creating
```
(EvaluationLink
   (PredicateNode "that")
   (ListLink
      (EvaluationLink
          (PredicateNode "glad@11")
          (ListLink
              (ConceptNode "he@33")))
      (EvaluationLink
          (PredicateNode "won@22")
          (ListLink
              (ConceptNode "she@44")))))
```
Is that not the original intention?

**Comment by anitzkin** [comment]
Yup, even as I was writing the last post, I realized that thematic roles
are too advanced to be determined within R2L,
but let's go back to what prompted the idea; I brought it up because of the
problem raised by ?William?'s illustration
of PredicateNodes and their EvaluationLinks -- the problem I see being that
the grammatical / semantic roles of the
arguments, vis-a-vis subject, object, indirect-object, other complement,
are unspecified.  Thematic roles would be one
solution to that problem, but perhaps one going too far.  Very simple
heuristics could bequeath the needed roles
temporarily, to be more certainly resolved along with the fuller sentential
semantics at a later stage.  The other things
that prompted the suggestion were:

-- that LG directly encodes some thematic roles, e.g. the "to" that means
"in order to" in "Bill called to say hello." has
its own link.  Which can then immediately assign its the object the
theta-role "goal"
-- that Relex partially attempts to infer things like this already, mainly
in terms of your half-implemented %relations which
I augmented last year as part of installing question functionality.
Although there's also the wealth of detail already installed in the
comparative relations.
-- that some of the Relex output was previously fairly inappropriate for
R2L to do anything with, such as the individual preposition
relations, and the incorrect conflation of complementizers with
prepositions, relativizers, and other things.  The difference between
a realtive clause, a sentential complement, and some other constituent
types definitely needs to be sorted out for the sake of R2L -- which, these
are the things I'm currently trying to fix, mainly.

So, I guess the point is that if we are talking about re-designing R2L, I
would like us to make some more principled
decisions about where semantic elements get differentiated -- by relex, by
R2L, or later -- some of the current
pipeline doing it too early or too late.  To me, this means that its way
past time to sketch out the semantic processing
layers of the pre-AGI dialogue system, including its all-important
interface with non-verbal cognition, which seems
a big question-mark at present?

**Comment by bgoertzel** [comment]
the semantic processing

> layers of the pre-AGI dialogue system, including its all-important
> interface with non-verbal cognition, which seems
> a big question-mark at present?



I think that the video game world provides a good context for thinking
about this in depth -- e.g. simple sentences about observations and actions
in the game world.   An advantage here is that we already have some code
that embodies choices made regarding Atomspace representation of
nonlinguistic relationships...

-- Ben

**Comment by anitzkin** [comment]
Yeah, in fact, we started a valuable discussion with Lake about just that, after lunch about a month ago, and I would want to discuss it in depth with someone intimately familiar with the gameworld code before sketching much . . . have you any specific intentions yet regarding how to pick up from where Lake left us?

**Comment by bgoertzel** [comment]
Lake will be resuming work on the game world as of late Feb (and thru the
end of June) ...

My general  intention re the game world work is to focus for a while on
fixing issues with Embodiment and replacing the XML messaging with ROS
messaging ... and once that is done return to the game world stuff....  My
current understanding is that what was bogging down the game world work was
not game world stuff per se, but stuff related to messaging and Embodiment
hackiness...

-- Ben

On Tue, Jan 13, 2015 at 10:08 PM, anitzkin <notifications@github.com> wrote:

> Yeah, in fact, we started a valuable discussion with Lake about just that,
> after lunch about a month ago, and I would want to discuss it in depth with
> someone intimately familiar with the gameworld code before sketching much .
> . . have you any specific intentions yet regarding how to pick up from
> where Lake left us?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1213#issuecomment-69866436>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by williampma** [comment]
So.... sorry to interrupt, but is adding a PLN rule to replace `PredicateNode` not in the 1st slot of an `EvaluationLink`, with an `EvaluationLink` with the `PredicateNode` is in the 1st slot, a correct logical inference?

Something like saying that "if P then Q" then "if P(x) then Q(x)"...

**Comment by bgoertzel** [comment]
William,

You said

***
So, with the example in the original post, PLN would end up creating

(EvaluationLink
   (PredicateNode "that")
   (ListLink
      (EvaluationLink
          (PredicateNode "glad@11")
          (ListLink
              (ConceptNode "he@33")))
      (EvaluationLink
          (PredicateNode "won@22")
          (ListLink
              (ConceptNode "she@44")))))

****

I just spent about 5 minutes poking around in gmail and github trying to
find "the example in the original post" but without success ;p

As I understand it, you want the above snippet to be the *output* of a
certain PLN inference...   Could you remind me what you want to be the
*input* of that inference?

Thanks
Ben

On Tue, Jan 13, 2015 at 10:31 PM, William Ma <notifications@github.com>
wrote:

> So.... sorry to interrupt, but is adding a PLN rule to replace
> PredicateNode not in the 1st slot of an EvaluationLink, with an
> EvaluationLink with the PredicateNode is in the 1st slot, a correct
> logical inference?
>
> Something like saying that "if P then Q" then "if P(x) then Q(x)"...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1213#issuecomment-69867871>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by williampma** [comment]
Hello Ben,

Here's the original input:
```
(EvaluationLink
   (PredicateNode "that")
   (ListLink
      (PredicateNode "glad@11")
      (PredicateNode "won@22")))
(EvaluationLink
   (PredicateNode "glad@11")
   (ListLink
       (ConceptNode "he@33")))
(EvaluationLink
   (PredicateNode "won@22")
   (ListLink
       (ConceptNode "she@44")))
```

**Comment by bgoertzel** [comment]
Hmmm....   That's subtle.

The input you give does not, in itself, entail the output you want.

Because, what if there were also other links such as

(EvaluationLink
   (PredicateNode "glad@11")
   (ListLink
       (ConceptNode "frog@88")))

in the Atomspace?

Then reasoning from your input to your output wouldn't really work...

The inference you suggest seems to contain the assumption that the relations

(EvaluationLink
   (PredicateNode "glad@11")
   (ListLink
       (ConceptNode "he@33")))
(EvaluationLink
   (PredicateNode "won@22")
   (ListLink
       (ConceptNode "she@44")))

summarize all that is known about the evaluation of the predicates
glad@11 and won@22

It seems what we have here, though, is the relation between the abstract
definition of a function as a symbol

F

and the description of a function as a set of ordered pairs

F_list = (x1, F(x1) ),  ... , (xN, F(xN) )

In your particular example, N=1

If we have a function

h(F)

we should be able to define an associated function h* so that

h*( F_list) = h(F)

The function h* is semantically basically the same as h, but mathematically
not identical (it has a different argument type).    Getting all
Haskell-ish, we
could write

h* = (Bourbakify h)

http://en.wikipedia.org/wiki/History_of_the_function_concept#Bourbaki_1939

where Bourbakify is a combinator mapping h into h* for any function

We could then say, in your example, that if we have

(EvaluationLink
   (PredicateNode "that")
   (ListLink
      (PredicateNode "glad@11")
      (PredicateNode "won@22")))
(EvaluationLink
   (PredicateNode "glad@11")
   (ListLink
       (ConceptNode "he@33")))
(EvaluationLink
   (PredicateNode "won@22")
   (ListLink
       (ConceptNode "she@44")))

and the latter two relations are the only Evaluation relations known
regarding glad@11 and won@22, then we have

(EvaluationLink
   (PredicateNode "that*")
   (ListLink
      (EvaluationLink
          (PredicateNode "glad@11")
          (ListLink
              (ConceptNode "he@33")))
      (EvaluationLink
          (PredicateNode "won@22")
          (ListLink
              (ConceptNode "she@44")))))


or equivalently

(EvaluationLink
   (ExecutionOutputLink

       (PredicateNode "Bourbakify")

       (PredicateNode "that")
   )
   (ListLink
      (EvaluationLink
          (PredicateNode "glad@11")
          (ListLink
              (ConceptNode "he@33")))
      (EvaluationLink
          (PredicateNode "won@22")
          (ListLink
              (ConceptNode "she@44")))))


Of course I just made up the ugly term "Bourbakify"; maybe some
functional programming or combinatory logic formalism has another
nicer name for it...

The above plays fast and loose with the distinction between elements
and single-element lists, but hopefully you get the idea...

The key point is, the "that" in your input and the "that" in your output
aren't really easily thought of as mathematically the same, though they
can be thought of as predicates that act on mathematical rearrangements
of each others' arguments...

So we could make a rule that does what you want if we are clear to
distinguish
"that" from "that*" or whatever we want to call it ....  Of course, NL
generation
and other crude linguistic methods can feel free to ignore the distinction
between
"that" and "that*", but PLN can't really ignore the distinction without
drawing
problematic inferences, IMO...

-- Ben

p.s. this stuff is a bit complex, I had to sit and stare and wonder for
about 4.32716 minutes before writing this email...

On Wed, Jan 14, 2015 at 12:33 AM, William Ma <notifications@github.com>
wrote:

> Hello Ben,
>
> Here's the original input:
>
> (EvaluationLink
>    (PredicateNode "that")
>    (ListLink
>       (PredicateNode "glad@11")
>       (PredicateNode "won@22")))
> (EvaluationLink
>    (PredicateNode "glad@11")
>    (ListLink
>        (ConceptNode "he@33")))
> (EvaluationLink
>    (PredicateNode "won@22")
>    (ListLink
>        (ConceptNode "she@44")))
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1213#issuecomment-69875004>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by williampma** [comment]
Humm... I guess it is true that "that" and "that*" are not exactly the same, but I am not sure how that would make PLN draws incorrect inferences.  Even with the extra "frog@88" link, it would still be OK for PLN to create something like
```
(EvaluationLink
    (PredicateNode "that")
    (ListLink
        (SetLink/MemberLink/OrLink or something....
            (EvaluationLink
                (PredicateNode "glad@11")
                (ListLink
                    (ConceptNode "he@33")))
            (EvaluationLink
                (PredicateNode "glad@11")
                (ListLink
                    (ConceptNode "frog@88))))
    (EvaluationLink (PredicateNode "won@22") (ListLink (ConceptNode "she@44")))))
```
since the original link never implies otherwise.  Maybe it will take some non-linguistic context to see how it will make incorrect inferences.

And if this is not done by PLN, then I am completely confused what is the exact algorithmic steps between R2L's current output, to NLU and to NLG all the way to surface realization.

**Comment by bgoertzel** [comment]
William,

Hmm, i thought about it more and I guess you're right that it won't cause
incorrect inferences.  I guess it may be more like "operator
overloading."   In other words, we are overloading (PredicateNode "that")
by using it to refer to two different mathematical functions, that have
related conceptual meanings but different argument types.

This may be OK to do, however, PLN is not going to do it automatically,
because the input you're providing and the output you want to produce are
NOT logically equivalent....   Instead, you'd have to create a special
"macro rule" to do this.   PLN could then apply this special "macro rule",
executing the kind of predicate overloading you're suggesting...

After a night's sleep, it seems OK to me to create a special rule that does
the transformation you want, and have PLN apply this rule....   I think the
rule should look like

EvaluationLink
     PredicateNode F
     PredicateNode P

|-

EvaluationLink
     PredicateNode F
     ANDLink
         EvaluationLink
             PredicateNode P
             Atom Arg1
         EvaluationLink
             PredicateNode P
             Atom Arg2
          ...

...

and should also cover the case where instead of just

PredicateNode P

we have

LIst
    PredicateNode P
    PredicateNode Q

or longer lists of this nature...

I suppose this is what one would call an "abuse of notation" in a math
textbook; but abuses of notation are often indispensable for actually doing
math...

-- Ben

p.s. I'll be offline for about the next 36 hrs or so, then check email
again late HK Friday morning...


On Wed, Jan 14, 2015 at 7:42 AM, William Ma <notifications@github.com>
wrote:

> Humm... I guess it is true that "that" and "that*" are not exactly the
> same, but I am not sure how that would make PLN draws incorrect inferences.
> Even with the extra "frog@88" link, it would still be OK for PLN to
> create something like
>
> (EvaluationLink
>     (PredicateNode "that")
>     (ListLink
>         (SetLink/MemberLink/OrLink or something....
>             (EvaluationLink
>                 (PredicateNode "glad@11")
>                 (ListLink
>                     (ConceptNode "he@33")))
>             (EvaluationLink
>                 (PredicateNode "glad@11")
>                 (ListLink
>                     (ConceptNode "frog@88))))
>     (EvaluationLink (PredicateNode "won@22") (ListLink (ConceptNode "she@44")))))
>
> since the original link never implies otherwise. Maybe it will take some
> non-linguistic context to see how it will make incorrect inferences.
>
> And if this is not done by PLN, then I am completely confused what is the
> exact algorithmic steps between R2L's current output, to NLU and to NLG all
> the way to surface realization.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1213#issuecomment-69917121>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by linas** [comment]
Aaron:
> I suggest that we try a principled, fore-thought-out version! 

Yes, no complaint there. I was assuming that this was a very narrow-issue PLN-specific bug report, not an anything-goes design discussion, which is what its evolved into.  GITHUB bug reports are not the best places to have wide-ranging design discussions, 

> Without some smart sense disambiguation, all we can do is associate an instance of a verb with a list of *potential* semantic role labels for its arguments... 

I propose that a sense-disambiguation framework be started, and nominate both the mihalcea code, and the anaphora-resolution code as the first plugins.  In a sense, anaphora-resolution is a special-case for sense-disambiguation, working at the level of pronoun referents.

**Comment by linas** [comment]
Aaron:
> ... the current pipeline doing it too early or too late ...
> ... its way past time to sketch out the semantic processing layers of the pre-AGI dialogue system...

Yes, but please be aware that this directly impacts/interacts-with/conflicts with language learning. The very first steps of language learning will extract link-grammar-like links (by design, that is explicitly what it tries to do), and I fully expect these to resemble the current dictionaries.  In particular, that means that semantic and syntactic info will be tangled together, just as it is now. 

The current learning proposal is only very concrete w.r.t. learning LG link types.  How to learn higher-layer stuff is vaguer, and would benefit from/be-influenced by a sketch of the semantic processing layers.

**Comment by anitzkin** [comment]
Yes, I'm aware of the conflict with unsupervised language-learning.  I was
speaking from the perspective of continuing to
work on a partially hand-coded version of the dialogue system because I
still consider our currently planned approach to
 unsupervised learning to be a long-shot.

I agree that the best that we can do in terms of setting up an accurate
sub-structure for inferring semantic roles is listing
potential roles.  The meat of the role-assignment has to be coded in terms
of the criteria attached to the roles themselves,
not the slots. But evaluating those criteria require the existence of
semantic structure informing the predicates and the arguments,
for which there are no provisions in the system yet . . .
